### PR TITLE
Convert to pure WebExtensions

### PR DESCRIPTION
**Pull request by jobisoft**
This PR removes your Experiment and replaces it by WebExtension code.

To get the body, I use `messages.getFull()` and search for the first `text/plain` body part. I included a 3rd party library to parse email addresses. To make things simple, I converted the background to a module to be able to use ES6 modules.

### CODE REVIEW

**Comment by jobisoft** [comment]
The onCommand event now has a second parameter to give you the active tab, so I could remove `getCurrentWindow()`.

File: manifest.json [hunk 1]
```diff
@@ -3,23 +3,20 @@
   "applications": {
     "gecko": {
       "id": "copypatch@kiszka.org",
-      "strict_min_version": "78.4.0",
-      "strict_max_version": "102.*"
+      "strict_min_version": "102.0"
```
Comment (jan-kiszka) [inline manifest.json:6]: Why don't we need a `strict_max_version` anymore?

File: manifest.json [hunk 2]
```diff
@@ -3,23 +3,20 @@
...
     }
   },
   "name": "Copy Patch",
   "description": "Copy email content to clipboard for application as patch",
   "author": "Jan Kiszka",
-  "version": "2.1.3",
+  "version": "2.1.4",
```
Comment (jan-kiszka) [inline manifest.json:12]: Release will happen later.

File: email-address-parser/email-addresses.mjs [hunk 3]
```diff
@@ -0,0 +1,1087 @@
...
+        }
+    }
+    return opts;
+}
+
+parse5322.parseOneAddress = parseOneAddressSimple;
+parse5322.parseAddressList = parseAddressListSimple;
+parse5322.parseFrom = parseFromSimple;
+parse5322.parseSender = parseSenderSimple;
+parse5322.parseReplyTo = parseReplyToSimple;
```
Comment (jan-kiszka) [inline email-address-parser/email-addresses.mjs:1087]: This looks like a valuable tool, but I'm not a fan of forking it here. We should probably make this addon an npm package so that its dependency can be pulled with standard tools.

**Comment by jan-kiszka** [comment]
Thanks for the conversion. The changes look helpful, will test them in a minute.

However, I would rather like to do the conversion in smaller, logically separated commits. If you have the time, please factor out e.g. style changes, getCurrentWindow removal etc. Let me know if you have time to do that as well or if I should help out.

File: manifest.json (hunk 1 above)
Comment (jobisoft) [reply manifest.json:6]: This is a pure WebExtension, the Thunderbird Team makes sure it stays compatible, because they keep the APIs stable/working.

File: email-address-parser/email-addresses.mjs (hunk 3 above)
Comment (jobisoft) [reply email-address-parser/email-addresses.mjs:1087]: If you have the bandwidth to do that, this is of course doable. I did not want to force you to having to build your add-on using a complex toolchain.

**Comment by jobisoft** [comment]
What style changes are you referring to? All I did was to move your legacy code into your background and include a 3rd party library. I do not see how that can be done in smaller steps and having a working add-on after each step. What part are you worried about in particular? I could add more explanatory comments, if that would help, but I do not have the bandwidth to redo this PR.

I have many other add-ons, where I have to support developers to move from legacy to WebExt: https://thundernest.github.io/add-on-reports/lost-tb102-to-tb115-worst-case.html

Thunderbird will move to stable monthly releases in 2024, and it is not plausible to expect add-on devs to release a new version for their Experiment add-ons every 4 weeks. So we have to get them away from Experiments where possible.

File: background-script.js [hunk 4]
```diff
@@ -11,62 +11,119 @@
...
             messenger.messageDisplayAction.setBadgeBackgroundColor(
-                {tabId: sender.tab.id, color: "green"});
+                { tabId: sender.tab.id, color: "green" });
             messenger.messageDisplayAction.setBadgeText(
-                {tabId: sender.tab.id, text: "✔"});
+                { tabId: sender.tab.id, text: "✔" });
             setTimeout(() => {
                 messenger.messageDisplayAction.setBadgeText(
-                    {tabId: sender.tab.id, text: null});
+                    { tabId: sender.tab.id, text: null });
```
Comment (jan-kiszka) [inline background-script.js]: These three are examples for style changes I was referring to.

**Comment by jan-kiszka** [comment]
No problem, then I will rework the changes myself.

BTW, there is also https://addons.thunderbird.net/de/thunderbird/addon/toggle-line-wrap which is still using Experiments because TB lacks that feature natively (https://bugzilla.mozilla.org/show_bug.cgi?id=1587153). Just had to bump its version as well, and if I have to do that every other month soon, that would become rather unhandy.

File: background-script.js (hunk 4 above)
Comment (jobisoft) [reply background-script.js]: I will try to undo the automatic pretty format changes caused by Visual Studio Code tomorrow.

**Comment by jan-kiszka** [comment]
Don't worry, I will rebuild the series to form a clean patch queue that can go into master.

**Comment by jobisoft** [comment]
I think toggle line wrap can be recreated as a pure WebExtension without using any internal Thunderbird stuff. Either by using a compose script to inject a css to toggle a line wrap visually and use onBeforeSend event to actually manipulate the outgoing plaintext mail to break at the desired places, or manipulate the DOM of the message while the user is typing.

**Comment by jan-kiszka** [comment]
> I think toggle line wrap can be recreated as a pure WebExtension without using any internal Thunderbird stuff. Either by using a compose script to inject a css to toggle a line wrap visually and use onBeforeSend event to actually manipulate the outgoing plaintext mail to break at the desired places, or manipulate the DOM of the message while the user is typing.

Honestly, I would rather prefer this finally being addressed in TB because none of those solutions seems simple to me. So, if I have to invest time, that should go into a upstream patch. Let's follow up on that issue.

File: manifest.json [hunk 5]
```diff
@@ -3,23 +3,20 @@
...
   "homepage_url": "http://git.kiszka.org/copypatch.git",
   "icons": {
     "32": "copypatch.png",
     "64": "copypatch64.png"
   },
   "background": {
-    "scripts": [
-      "background-script.js"
-    ]
+    "page": "background.html"
```
Comment (jan-kiszka) [inline manifest.json:19]: What's the advantage of this change?

Comment (jobisoft) [reply manifest.json:19]: You can use ES6 modules.

Comment (jobisoft) [reply manifest.json:19]: https://developer.thunderbird.net/add-ons/mailextensions#background-page

Comment (jan-kiszka) [reply manifest.json:19]: Ok, that allows the import later on - understood. But it causes this warning: `This page is in Quirks Mode. Page layout may be impacted. For Standards Mode use “<!DOCTYPE html>”.`

File: manifest.json (hunk 1 above)
Comment (jan-kiszka) [inline manifest.json:6]: Must become 106.0 due to using the tab parameter of the onCommand callback.

File: manifest.json (hunk 5 above)
Comment (jobisoft) [reply manifest.json:19]: I did add that line to background.html in my PR and I do not see the error here. Are you 100% sure it is caused by copy patch and not by another html page loaded somewhere?

File: manifest.json (hunk 1 above)
Comment (jobisoft) [reply manifest.json:6]: It was backported to 102:
https://searchfox.org/comm-esr102/source/mail/components/extensions/parent/ext-commands.js#46-49

You sure it does not work?

Comment (jobisoft) [reply manifest.json:6]: It is documented in the 102 branch of the API as well:
https://webextension-api.thunderbird.net/en/102/commands.html#events

Comment (jan-kiszka) [reply manifest.json:6]: I was reading https://webextension-api.thunderbird.net/en/latest/commands.html - seems your documentation is outdated then. To which version of 102 was this added then exactly?

Comment (jan-kiszka) [reply manifest.json:6]: Ah, 102.3.3 then.

Comment (jobisoft) [reply manifest.json:6]: But to be precise, it must be set to 102.4.0. So indeed a good catch! (102.3.3 does not exist in the list of [allowed versions](https://addons.thunderbird.net/pages/appversions/))

File: background-script.js [hunk 6]
```diff
@@ -40,33 +99,30 @@ function main()
                 messenger.messageDisplayAction.setBadgeText(
                     {tabId: sender.tab.id, text: null});
             }, 500);
+            return Promise.resolve();
         }
+        return false;
```
Comment (jan-kiszka) [inline background-script.js:104]: Factoring out these two lines which are not directly related to the Experiment API conversion: Why is it correct to just call `Promise.resolve()` without any `value` parameter? I'm not reading that case out of its documentation but would still like to write a good commit message.

File: manifest.json (hunk 5 above)
Comment (jan-kiszka) [reply manifest.json:19]: I was caused by the missing update of the Makefile, not packaging `background.html` yet.

File: email-address-parser/email-addresses.mjs [hunk 7]
```diff
@@ -0,0 +1,1087 @@
...
+//         }, ...]
+// }
+//
+// emailAddresses.parseOneAddress and emailAddresses.parseAddressList
+// work as you might expect. Try it out.
+//
+// Many thanks to Dominic Sayers and his documentation on the is_email function,
+// http://code.google.com/p/isemail/ , which helped greatly in writing this parser.
+
+export function parse5322(opts) {
```
Comment (jan-kiszka) [inline email-address-parser/email-addresses.mjs:28]: Meanwhile I understood that this was a local modification on top of the original version. But we can do better, we can go upstream first (provided upstream is still merging): https://github.com/jackbearheart/email-addresses/pull/68

File: background-script.js (hunk 6 above)
Comment (jan-kiszka) [reply background-script.js:104]: Ok, found on stackoverflow: empty value means undefined value, and that is apparently a valid usage.

File: background-script.js [hunk 8]
```diff
@@ -11,23 +11,82 @@
...
+        if (part.parts) {
+            let body = getBody(part.parts);
+            if (body) {
+                return body
+            }
+        }
     }
     return null;
 }
+// Compatibility layer for former experiment.
```
Comment (jan-kiszka) [inline background-script.js:56]: Reads like as if there could be a better way. However, to my understanding, there is no way for the content script to access `browser.messages` directly. Therefore, we need such an interface, no?

**Comment by jan-kiszka** [comment]
#2 is supposed to carry everything of this PR, just broken up and slightly massaged.
